# DATASET

In [2]:
# imports
import pandas as pd

I have a few disjointed excel spreadsheets. I have three separate spreadsheets and one PDF that contain the data I need.

## Monthly Sales

One is called "Sales by part of the month" and turned out to be a pivot table. It contains the following info:
* Row labels contain the product SKUs (not uniform numbers). They are grouped together by customer categories, which are uniform six-figure numbers (except for one!). 
* Column labels go from 2012-04 to 2017-10, monthly sales data.
Data is sparse (demand is intermittent). 

To-do items:
* READ IN second tab to create df with time series data.

In [11]:
monthly = pd.read_excel('Sales by part by month.xlsx')
monthly.head(20)

Unnamed: 0     Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  \
0   Sum of Final Price  Column Labels        NaN        NaN        NaN   
1           Row Labels        2012-04    2012-05    2012-06    2012-07   
2               400000            NaN        NaN        NaN        NaN   
3          0300ST365-1            NaN        NaN        NaN        NaN   
4            0311HT265            NaN        NaN        NaN        NaN   
5            0311HT320            NaN        NaN        NaN        NaN   
6           03139731BT            NaN        NaN        NaN        NaN   
7          03139800MPK            NaN        NaN        NaN        NaN   
8            0313FC320            NaN        NaN        NaN        NaN   
9            0313FC365            NaN        NaN        NaN        NaN   
10          03149731CC            NaN        NaN        NaN        NaN   
11          03149731FT            NaN        NaN        NaN        NaN   
12       03149800MIN-1            NaN        NaN        NaN        NaN   
13         0314FT320-1            NaN        NaN        NaN        NaN   
14         0314FT365-2            NaN        NaN        NaN        NaN   
15         03159800-14            NaN        NaN        NaN        NaN   
16          03159800-8            NaN        NaN        NaN        NaN   
17         03159800MBX            NaN        NaN        NaN        NaN   
18             1814/23            NaN        NaN        NaN        NaN   
19                2006            NaN        NaN        NaN        NaN   

   Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9     ...       \
0         NaN        NaN        NaN        NaN        NaN     ...        
1     2012-08    2012-09    2012-10    2012-11    2012-12     ...        
2         NaN        NaN        NaN        NaN        NaN     ...        
3         NaN        NaN        NaN        NaN        NaN     ...        
4         NaN        NaN        NaN        NaN        NaN     ...        
5         NaN        NaN        NaN        NaN        NaN     ...        
6         NaN        NaN        NaN        NaN        NaN     ...        
7         NaN        NaN        NaN        NaN        NaN     ...        
8         NaN        NaN        NaN        NaN        NaN     ...        
9         NaN        NaN        NaN        NaN        NaN     ...        
10        NaN        NaN        NaN        NaN        NaN     ...        
11        NaN        NaN        NaN        NaN        NaN     ...        
12        NaN        NaN        NaN        NaN        NaN     ...        
13        NaN        NaN        NaN        NaN        NaN     ...        
14        NaN        NaN        NaN        NaN        NaN     ...        
15        NaN        NaN        NaN        NaN        NaN     ...        
16        NaN        NaN        NaN        NaN        NaN     ...        
17        NaN        NaN        NaN        NaN        NaN     ...        
18        NaN        NaN        NaN        NaN        NaN     ...        
19        NaN        NaN        NaN        NaN        NaN     ...        

   Unnamed: 60 Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65  \
0          NaN         NaN         NaN         NaN         NaN         NaN   
1      2017-03     2017-04     2017-05     2017-06     2017-07     2017-08   
2            0           0           0           0           0           0   
3          NaN         NaN         NaN         NaN         NaN         NaN   
4          NaN         NaN         NaN         NaN         NaN         NaN   
5          NaN         NaN         NaN         NaN         NaN         NaN   
6          NaN         NaN         NaN         NaN         NaN         NaN   
7          NaN         NaN         NaN         NaN         NaN         NaN   
8          NaN         NaN         NaN         NaN         NaN         NaN   
9          NaN         NaN         NaN         NaN         NaN         NaN   
10         NaN         NaN         NaN      

In [18]:
monthly.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128786 entries, 0 to 128785
Data columns (total 70 columns):
Unnamed: 0     128786 non-null object
Unnamed: 1     5714 non-null object
Unnamed: 2     5773 non-null object
Unnamed: 3     6206 non-null object
Unnamed: 4     6876 non-null object
Unnamed: 5     7399 non-null object
Unnamed: 6     8822 non-null object
Unnamed: 7     9056 non-null object
Unnamed: 8     8419 non-null object
Unnamed: 9     7824 non-null object
Unnamed: 10    8266 non-null object
Unnamed: 11    6946 non-null object
Unnamed: 12    7799 non-null object
Unnamed: 13    6492 non-null object
Unnamed: 14    6887 non-null object
Unnamed: 15    6564 non-null object
Unnamed: 16    7171 non-null object
Unnamed: 17    7976 non-null object
Unnamed: 18    9282 non-null object
Unnamed: 19    11214 non-null object
Unnamed: 20    8021 non-null object
Unnamed: 21    9108 non-null object
Unnamed: 22    7754 non-null object
Unnamed: 23    6974 non-null object
Unnamed: 24    6405 no

In [3]:
monthly = pd.read_excel('Sales by part by month.xlsx', sheet_name='Sheet3')
monthly.head(20)

Part  Unit Price  Quantity  Customer     InvNo  Final Price    EntDate  \
0    8812-2         0.0         4    409060  14800000          0.0 2012-04-03   
1    7267-2        30.0        -1    404230  14800001        -30.0 2012-04-04   
2    2922-7         0.0         8    408130  14800082          0.0 2012-04-06   
3    2944-7        12.0         6    403980  14800071         72.0 2012-04-06   
4    2944-7        15.0         4    403710  14800072         60.0 2012-04-06   
5      4472        40.0         2    400590  14800051         80.0 2012-04-06   
6    4472-7        40.0         1    402390  14800093         40.0 2012-04-06   
7      4476        40.0         1    400590  14800051         40.0 2012-04-06   
8      4483        31.9         2    403830  14800085         63.8 2012-04-06   
9      8861        13.0         3    407510  14800053         39.0 2012-04-06   
10   7250-1         0.0         1    413100  14800015          0.0 2012-04-06   
11   7267-1        20.0         1    403710  14800072         20.0 2012-04-06   
12   7267-3         0.0         1    406900  14800002          0.0 2012-04-06   
13     8000        10.5         1    404150  14800055         10.5 2012-04-06   
14     8000        10.5         5    407510  14800053         52.5 2012-04-06   
15  7372/12         0.0         8    408130  14800082          0.0 2012-04-06   
16   7372/8        16.0         1    400590  14800051         16.0 2012-04-06   
17   7372/8         0.0         2    413040  14800058          0.0 2012-04-06   
18  7379/10        36.0         6    404900  14800057        216.0 2012-04-06   
19   5558-1        10.0         6    407130  14800052         60.0 2012-04-06   

      PLn  Agt    InvDate      RecordID  InvDate.1 InvDate 2 InvDate3  
0   50612   70 2012-04-03  R|51236|0|10 2012-04-03  Apr-2012  2012-04  
1   20202   61 2012-04-04  R|51237|0|10 2012-04-04  Apr-2012  2012-04  
2   20008   99 2012-04-06  R|51319|0|10 2012-04-06  Apr-2012  2012-04  
3   20008   15 2012-04-06  R|51308|0|10 2012-04-06  Apr-2012  2012-04  
4   20008   15 2012-04-06  R|51309|0|10 2012-04-06  Apr-2012  2012-04  
5   20008   20 2012-04-06  R|51288|0|50 2012-04-06  Apr-2012  2012-04  
6   20008   65 2012-04-06  R|51330|0|90 2012-04-06  Apr-2012  2012-04  
7   20008   22 2012-04-06  R|51288|0|60 2012-04-06  Apr-2012  2012-04  
8   20008    6 2012-04-06  R|51322|0|20 2012-04-06  Apr-2012  2012-04  
9   20008   16 2012-04-06  R|51290|0|90 2012-04-06  Apr-2012  2012-04  
10  20202   61 2012-04-06  R|51252|0|10 2012-04-06  Apr-2012  2012-04  
11  20202   99 2012-04-06  R|51309|0|30 2012-04-06  Apr-2012  2012-04  
12  20202    6 2012-04-06  R|51238|0|50 2012-04-06  Apr-2012  2012-04  
13  20202   22 2012-04-06  R|51292|0|30 2012-04-06  Apr-2012  2012-04  
14  20202   22 2012-04-06  R|51290|0|80 2012-04-06  Apr-2012  2012-04  
15  20203   65 2012-04-06  R|51319|0|80 2012-04-06  Apr-2012  2012-04  
16  20203   70 2012-04-06  R|51288|0|80 2012-04-06  Apr-2012  2012-04  
17  20203   70 2012-04-06  R|51295|0|20 2012-04-06  Apr-2012  2012-04  
18  20203   20 2012-04-06  R|51294|0|10 2012-04-06  Apr-2012  2012-04  
19  20301   99 2012-04-06  R|51289|0|30 2012-04-06  Apr-2012  2012-04

In [4]:
monthly.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 922077 entries, 0 to 922076
Data columns (total 14 columns):
Part           922077 non-null object
Unit Price     922077 non-null float64
Quantity       922077 non-null int64
Customer       922077 non-null int64
InvNo          922077 non-null int64
Final Price    922077 non-null float64
EntDate        922077 non-null datetime64[ns]
PLn            922077 non-null int64
Agt            922077 non-null int64
InvDate        922077 non-null datetime64[ns]
RecordID       922077 non-null object
InvDate.1      922077 non-null datetime64[ns]
InvDate 2      922077 non-null object
InvDate3       922077 non-null object
dtypes: datetime64[ns](3), float64(2), int64(5), object(4)
memory usage: 98.5+ MB


In [6]:
monthly.describe()

Unit Price       Quantity       Customer         InvNo  \
count  922077.00000  922077.000000  922077.000000  9.220770e+05   
mean       48.27954      10.679287  405662.277386  1.438898e+07   
std       239.54981      59.570260    4080.905777  3.886682e+05   
min         0.00000   -1030.000000  400000.000000  1.400000e+07   
25%        11.00000       1.000000  402120.000000  1.406094e+07   
50%        30.00000       2.000000  404820.000000  1.412024e+07   
75%        55.00000       6.000000  407910.000000  1.484688e+07   
max    218185.40000   10000.000000  499999.000000  1.491743e+07   

         Final Price            PLn            Agt  
count  922077.000000  922077.000000  922077.000000  
mean      397.883230   45225.684241      40.027002  
std      2420.515708   11462.769806      32.747810  
min   -199699.500000   20008.000000       5.000000  
25%        30.000000   50101.000000      15.000000  
50%        80.000000   50112.000000      20.000000  
75%       184.680000   50707.000000      67.000000  
max    240000.000000   52112.000000      99.000000

QUESTIONS to Lee:
* What do minus quantities mean? Voided order?
* How are EntDate and InvDate different? Do you prefer one over the other?
* Items with dashes (e.g. 4472 and 4472-7) are different items?
* There are a few items that seem fat-fingered entries. Maximum unit price is 218,185??? and quantities interquartile range is 1-6 but minimum is 1030 and maximum is 10,000?

## Promotions

Then I have three documents that contain the info on promoted items (for some curious reason they're called "kickers"):
* Kickers 2015 - excel - ranges from 2015/02-2016/03: one "item info" column containing the product item (first number), then months are blocked out in colors!!! 46 rows total.
* Kickers 2016 - excel - ranges from 2016/02-2017/01: same setup, 28 rows. 
* Kickers 2017 - PDF - for entire 2017: this will be manual... <br>

To-do items:
* CHECK 2016/02 and 2016/03 since the two excels overlap in dates.
* ASK Lee for kickers data on 2012/04-2015/02.
* CREATE one df?

CONCLUSION: IGNORE for now